### PR TITLE
Support 'include_type_name' in RestGetIndicesAction

### PR DESCRIPTION
**Pull request by cbuescher**
This change adds support for the 'include_type_name' parameter for the
indices.get API. This parameter, which defaults to `false` starting in 7.0,
changes the response to not include the indices type names any longer.

If the parameter is set in the request, we additionally emit a deprecation
warning since using the parameter should be only temporarily necessary while
adapting to the new response format and we will remove it with the next major
version.

### CODE REVIEW

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml
Comment (cbuescher) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml]: Note: this can be removed after the parameter support has been backported to 6.x

**Comment by cbuescher** [comment]
@elasticmachine run the gradle build tests 1

**Comment by cbuescher** [comment]
@elasticmachine run the gradle build tests 1

**Comment by cbuescher** [comment]
@elasticmachine run the gradle build tests 1

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: I think this comment can be removed?

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: Small comment, maybe `boolean includeTypeName` would be clearer here?

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesAction.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesAction.java]: Small comment, this can be `INCLUDE_TYPE_NAME_PARAMETER`.

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml
Comment (jtibshirani) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml]: I guess this case will need to be split out when backporting to 6.x, since the default will be different.

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: To check I understand, are we only retaining this method so that we can test the serialization of `GetIndexResponse`? It looks like we do not support `include_type_name = true` in the get indices HLRC call.

**Review by jtibshirani — COMMENTED** [review]
Thanks @cbuescher, this looks like it's on a good track. My main questions are around the Java HLRC:
- I had assumed we would flatten `GetIndexResponse#mappings` to be of type `ImmutableOpenMap<String, MappingMetaData>`, to avoid nesting the mappings under the `_doc` key.
- When we backport to 6.x, are we planning to add note under 'breaking changes', and also a 'breaking' label to the PR? My understanding from reading this PR is that for the Java HLRC we will force `include_type_name = false`.

Finally, maybe a Java HLRC expert could take a look to make sure this breaking approach is okay.

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml
Comment (cbuescher) [reply rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml]: probably yes, not sure yet though...

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: Yes, at some point I was planning on adding an alternative get-method (something like getWithTypes) to the IndicesClient, but now I think we don't need it unless we want ppl. to be able to get back their legacy type names from 6.x indices from the Java-Objects returned by the HLRC. If we are fine with this not being possible anymore I think we don't need this.

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: I agree that this is not so critical, as users should stop using the type name to encode information. So we could keep things simple and not add another method `getWithTypes`.

**Comment by cbuescher** [comment]
>     * I had assumed we would flatten `GetIndexResponse#mappings` to be of type `ImmutableOpenMap<String, MappingMetaData>`, to avoid nesting the mappings under the `_doc` key.

If we want to do this, I would only do it in the getter in GetIndexResponse, but leave the internal `mappings` field as is. Otherwise we also need to touch serialization and possible how the mappings are filled in TransportGetIndexAction. I think these are internal refactorings that we should do later, maybe even only on 8.0 where we remove all this cruft.

Changing the Java-API of GetIndexResponse however would be breaking, at least in 7.0. I wouldn't backport this change to 6.x. It would also mean ppl. would be able to get at the "old" type names via the REST API but not e.g. using the old transport client, which might be considered odd.

**Comment by cbuescher** [comment]
>     My understanding from reading this PR is that for the Java HLRC we will force `include_type_name = false`.

So far I thought we could make the 6.x HLRC client force `include_type_name = true` to get back the legacy format. The 6.x client needs to be forward compatible, so they should be able to talk to both 6.x and 7.0 nodes. I think it would help people transitioning if they would not need to update any of their own application logic before moving to the 7.0 client which will force the new response behaviour and will swallow the type information.

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: The `_default_` mapping is a now-deprecated feature that let users define a base mapping whose entries would be added to all newly created document mapping types (https://www.elastic.co/guide/en/elasticsearch/reference/current/default-mapping.html). It's added in the same way other mappings are added, the only difference is that it uses the reserved name `_default_`. It looks like our approach has been that when `include_type_name = false`, we try to find a mapping **other than** `_default_` to return.

It looks like `MapperService.DEFAULT_MAPPING` contains this constant.

Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: Thanks for the explanation, I pushed a change, let me know what you think.

**Comment by cbuescher** [comment]
@jtibshirani thanks for the review, I pushed a round of changes and tried to respond to your two questions above. Let me know if these answers make sense or you have objections. If we cannot decide on the `GetIndexResponse#mappings` change for now, we can leave it out of this PR perhaps and discuss it in more details elsewhere? It should be easy to add if we agree on doing so later on as well.

File: server/src/test/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponseTests.java
Comment (jtibshirani) [inline server/src/test/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponseTests.java]: I find it a little confusing that we can try to add the same type `_doc` multiple times, if `typeCount > 1`. Maybe this method should only take one parameter `boolean randomTypeName`, and then depending on its value we can choose a random `typeCount` within the method?

File: server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java
Comment (jtibshirani) [inline server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java]: Maybe we should use `randomBoolean()` here to make sure we continue to test typed responses? Now that we've removed `GetIndexResponse.fromXContent(XContentParser parser, boolean includeTypeName)` we can't test xContent serialization for typed responses in the usual way, so we would need to skip that or write a dedicated test.

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml
Comment (cbuescher) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml]: This started failing after I made the changes you requested to `GetIndexResponse#toXContent`. Not sure what the expected behaviour should be here, the index doesn't contain any user-specified mapping but it looks like we printed an empty one in the response so far. Change the test or update the `toXContent` rendering, wdyt?

Comment (jtibshirani) [reply rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml]: It seems like it's simplest to change the `toXContent` to match the old behavior, as this will avoid BWC concerns.

**Review by jtibshirani — COMMENTED** [review]
A few more comments, it looks very close! Sounds good about revisiting the `mappings` HLRC change later. It would be good to have a high-bandwidth discussion about this in person.

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml
Comment (cbuescher) [reply rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml]: done

File: server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java]: The problem is that the default toXContent/fromXContent methods now only support the "new" format. While we can output the legacy response format to XContent when setting `include_type_name=true` we cannot parse it. I removed that parsing method upon your request earlier, its the one we would have needed if we would add a "getWithTypes" method to the HLRC IndicesClient.
The AbstractStreamableXContentTestCase is set up in a way that toXContent/fromXContent are forming a pair, and since we don't have a second parsing method we also cannot test the roundtrip for the "old" output any longer.

File: server/src/test/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponseTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponseTests.java]: > we can try to add the same type `_doc` multiple times

Those will be put into the same map entry in the `typeBuilder`though.

> Maybe this method should only take one parameter `boolean includeTypeName`, and then depending on its value we can choose a random `typeCount` within the method? If it's `false`, then we only allow 0 or 1 types, and if it's `true` we could have more.

Creating mapping responses for 0/1 type can be desired for and without random type name though at the moment. `GetMappingsResponseTests` currently tests both 0-3 types and random type names, while in `GetIndexResponse` I need only 0/1 type and cannot test with random names because we cannot send them over the rest layer anymore which is why I need "_doc" to make the equality comparison in the roundtrip tests work.

Comment (jtibshirani) [reply server/src/test/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponseTests.java]: Right, I was suggesting two cases based on a single parameter `randomTypeName`:
- If `randomTypeName == true`, then use a random type name and select 0 to 3 types.
- If `randomTypeName == false`, then only use `_doc` and select 0 or 1 types.

In any case, this is a very minor point, happy to do whatever you find cleanest.

File: server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java
Comment (jtibshirani) [reply server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java]: Right, we wouldn't be able to test xContent serialization, but I think it should still be possible to still test stream serialization, equals and hashCode, etc.? I was thinking maybe we could override `AbstractStreamableXContentTestCase#createXContextTestInstance` to create only a typeless instance, but allow `createTestInstance` to create both typed + untyped versions.

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: Instead of adding the special check `if (indexMappings.size() == 0) { ... }`, another way to structure this would be to pull this `if (mappings == null) { ... } else { ... }` check out of the loop.

**Review by jtibshirani — APPROVED** [review]
I left a few comments for your consideration, but they are minor and I don't think need another round of review.

**Comment by jtibshirani** [comment]
One last note I forgot: it would be good to add `include_type_name` to the REST documentation for 'get index', with a note as we do for other APIs like 'get mapping'.

File: server/src/test/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponseTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponseTests.java]: I'd rather clean that up when we remove all this temporary types/noTypes confusion alltogether.

File: server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java]: I'll look into it.

Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/action/admin/indices/get/GetIndexResponseTests.java]: Great, that seems to work nicely. I didn't know that we could overwrite that method.

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: done

**Comment by cbuescher** [comment]
@jtibshirani thanks for the last round of review and the thumbs up, I pushed another commit adressing you last comments. Waiting for green CI now before I merge.

File: server/src/test/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesActionTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesActionTests.java]: I did that intially but it created some problems that I don't remember exactly now. I can try again but would prefer to do this without holding up this PR. The test is there at least.

**Comment by cbuescher** [comment]
@elasticmachine run the gradle build tests 1

**Comment by cbuescher** [comment]
@elasticmachine run default distro tests

File: server/src/test/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesActionTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesActionTests.java]: This seems to work now, but I'd like to postpone the change since I'm getting to a green CI build ;-)
I will need to switch the "include_type_name" flag after backporting this PR to 6.x anyway, I will include this update there as well then. Hope thats okay for you.

**Pull request merged**
